pypesto/model_selection/model_selection.py

```diff
@@ -0,0 +1,906 @@
+
+class ModelSelectionProblem:
+    """
+
+    """
+    def __init__(self,
+                 row: Dict[str, float],
+                 petab_problem: petab.problem,
+                 valid: bool = True,
+                 autorun: bool = True,
+                 x_fixed_estimated: Set[str] = None
+    ):
+        """
+        Arguments
+        ---------
+        row:
+            A single row from the model specification file, in the format that
+            is returned by `ModelSelector.model_generator()`.
+        petab_problem:
+            A petab problem that includes the parameters defined in the model
+            specification file.
+        valid:
+            If `False`, the model will not be tested.
+        autorun:
+            If `False`, the model parameters will not be estimated. Allows
+            users to manually call pypesto.minimize with custom options, then
+            `set_result()`.
+        fixed_estimated:
+            Parameters that can be fixed to different values can be considered
+            estimated, as the "best" fixed parameter will be preferred. Note,
+            the preference is implemented as comparison of different models
+            with `ModelSelectionMethod.compare()`, unlike normal estimation,
+            which occurs within the same model with `pypesto.minimize`.
+        TODO: constraints
+        """
+        self.row = row
+        self.petab_problem = petab_problem
+        self.valid = valid
+
+        if x_fixed_estimated is None:
+            x_fixed_estimated = set()
+        else:
+            # TODO remove parameters that are zero
+            pass
+
+        self.model_ID = self.row[MODEL_ID]
+
+        #self.AIC = None
+        self.BIC = None
+
+        if self.valid:
+            self.pypesto_problem = row2problem(row, petab_problem)
+
+            # TODO warning/error if x_fixed_estimated is not a parameter ID in
+            # the PEtab parameter table. A warning is currently produced in
+            # `row2problem` above.
+            # Could move to a separate method that is only called when a
+            # criterion that requires the number of estimated parameters is
+            # called.
+            self.estimated = x_fixed_estimated | set(
+                self.petab_problem.parameter_df.query(f'{ESTIMATE} == 1').index
+            )
+            self.n_estimated = len(self.estimated)
+            # TODO for BIC. Could move to `calculate_BIC()` to save time.
+            self.n_edata = None
+
+            self.minimize_result = None
+
+            if autorun:
+                self.set_result(minimize(self.pypesto_problem))
+
+    def set_result(self, result: Result):
+        self.minimize_result = result
+        self.optimized_model = self.minimize_result.optimize_result.list[0]
+
+    @property
+    def AIC(self):
+        if '_AIC' not in dir(self):
+            self._AIC = calculate_AIC(self.n_estimated,
+                                      self.optimized_model.fval)
```

Comment:
usually all attributes should be set in the `__init__` method already. Here you can check for `None` instead.